single-kernel/lin_reg_coeff.cpp

```diff
@@ -0,0 +1,196 @@
+
+      cgh.parallel_for<class VecProductKernel<T>>(ndrange,
+        [=](cl::sycl::nd_item<1> item) 
+        {
+          size_t gid= item.get_global_linear_id();
+          intermediate_product[gid] = in1[gid] * in2[gid];
+        });
+    });
+  }
+
+T reduce(s::buffer<T, 1> &input_buf) {
+  auto array_size = args.problem_size;
+  auto wgroup_size = args.local_size;
+  // Not yet tested with more than 2
+  auto elements_per_thread = 2;
+
+  while (array_size!= 1) {
+    auto n_wgroups = (array_size + wgroup_size*elements_per_thread - 1)/(wgroup_size*elements_per_thread); // two threads per work item
+
+    args.device_queue.submit(
+      [&](cl::sycl::handler& cgh) {
+
+        auto global_mem = input_buf.template get_access<s::access::mode::read_write>(cgh);
+    
+        // local memory for reduction
+        auto local_mem = s::accessor <T, 1, s::access::mode::read_write, s::access::target::local> {s::range<1>(wgroup_size), cgh};
+        cl::sycl::nd_range<1> ndrange (n_wgroups*wgroup_size, wgroup_size);
+  
+        cgh.parallel_for<class VecReduceKernel<T>>(ndrange,
+        [=](cl::sycl::nd_item<1> item) 
+          {
+            size_t gid= item.get_global_linear_id();
+            size_t lid = item.get_local_linear_id();
+
+            // initialize local memory to 0
+            local_mem[lid] = 0; 
+
+            if ((elements_per_thread * gid) < array_size) {
+                local_mem[lid] = global_mem[elements_per_thread*gid] + global_mem[elements_per_thread*gid + 1];
+            }
+
+            item.barrier(s::access::fence_space::local_space);
+
+            for (size_t stride = 1; stride < wgroup_size; stride *= elements_per_thread) {
+              auto local_mem_index = elements_per_thread * stride * lid;
+              if (local_mem_index < wgroup_size) {
+                  local_mem[local_mem_index] = local_mem[local_mem_index] + local_mem[local_mem_index + stride];
+              }
+
+              item.barrier(s::access::fence_space::local_space);
+            }
+
+            // Only one work-item per work group writes to global memory 
+            if (lid == 0) {
+              global_mem[item.get_group_linear_id()] = local_mem[0];
+            }
+          });
+      });
+    array_size = n_wgroups;
+  }
+  auto reduced_value = input_buf.template get_access<s::access::mode::read>();
+  return(reduced_value[0]);
+}
+
+  void run() {    
+    s::buffer<T, 1> input1_buf(input1.data(), s::range<1>(args.problem_size));
+    s::buffer<T, 1> input2_buf(input2.data(), s::range<1>(args.problem_size));
+    s::buffer<T, 1> output_buf(output.data(), s::range<1>(args.problem_size));
+
+    vec_product(input1_buf, input2_buf, output_buf);
+
+    T ss_xy = reduce(output_buf);
+
+    vec_product(input1_buf, input1_buf, output_buf);
+
+    T ss_xx = reduce(output_buf);
+
+    T mean_x = reduce(input1_buf)/args.problem_size;
+    T mean_y = reduce(input2_buf)/args.problem_size;
```

Comment:
Are we able to merge some of the kernels here for efficiency?

Reply:
Probably possible, but I am not sure the effect on caching, etc. because a thread needs to make access to two different arrays. Maybe AoS can help. But it requires more effort :)